Experimental Alertmanager API

gotjosh opened the pull request:
**What this PR does**:

Similarly to the ruler, this PR introduces a "configuration" API for the Alertmanager. It is per tenant, accepts/takes YAML and is backed up by object storage.

While I cannot take credit for this PR (as it was done by @jtlisi), I'm happy to follow up on reviews to see if I can get this merged.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

jtlisi commented on pkg/alertmanager/alerts/compat.go:
```diff
@@ -0,0 +1,32 @@
+	ErrNotFound = errors.New("alertmanager config not found")
+)
+
+// ToProto transforms a yaml Alertmanager config and map of template files to an AlertConficDesc
```
```suggestion
// ToProto transforms a yaml Alertmanager config and map of template files to an AlertConfigDesc
```

jtlisi commented on integration/configs.go:
```diff
@@ -51,9 +51,20 @@ var (
 	AlertmanagerFlags = map[string]string{
-		"-alertmanager.storage.local.path": filepath.Join(e2e.ContainerSharedDir, "alertmanager_configs"),
+		"-alertmanager.configs.poll-interval": "1s",
+		"-alertmanager.web.external-url":      "http://localhost/api/prom",
```
```suggestion
		"-alertmanager.web.external-url":      "http://localhost/alertmanager",
```

It might be worth updating this to use the URL that is active in both the `alertmanager` target and the `all` target

gotjosh commented on integration/e2ecortex/client.go:
```diff
@@ -173,18 +178,22 @@ type ServerStatus struct {
 // GetAlertmanagerConfig gets the status of an alertmanager instance
 func (c *Client) GetAlertmanagerConfig(ctx context.Context) (*alertConfig.Config, error) {
-	u := c.alertmanagerClient.URL("/api/v1/status", nil)
+	u := c.alertmanagerClient.URL("api/prom/api/v1/status", nil)
```
See my comment below for an explanation, this is right 😄

gotjosh commented on integration/e2ecortex/client.go:
```diff
@@ -284,3 +293,67 @@ func (c *Client) DeleteRuleGroup(namespace string, groupName string) error {
+
+// SetAlertmanagerConfig gets the status of an alertmanager instance
+func (c *Client) SetAlertmanagerConfig(ctx context.Context, amConfig string, templates map[string]string) error {
+	u := c.alertmanagerClient.URL("/api/v1/alerts", nil)
```
The `api/prom` prefix is used to list all the Prometheus compatible APIs under one place. Given that status of alerts which are firing comes from Prometheus this is listed under that prefix. The non-prefixed version is for the Cortex internal APIs which are not compatible.

gotjosh commented:
@tomwilkie @pracucci @gouthamve @jtlisi I have addressed the feedback given here as part of https://github.com/cortexproject/cortex/pull/2768/commits/a820a517ab1c8c4a72f2af41bd939aa26cdf39a7. I have also rebased to get rid of the changelog conflict.

Given all the comments so far have been about writing I think a review of the actual API is in place? 😄  Please do let me know if there's anything I can do to ease the reviewing process.

pracucci commented on integration/e2ecortex/client.go:
```diff
@@ -173,18 +178,22 @@ type ServerStatus struct {
 	}
 
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, fmt.Errorf("not found")
```
I guess you wanted to return `ErrNotFound`

pracucci commented on integration/e2ecortex/client.go:
```diff
@@ -284,3 +293,67 @@ func (c *Client) DeleteRuleGroup(namespace string, groupName string) error {
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return fmt.Errorf("not found")
```
`ErrNotFound`

pracucci commented on integration/e2ecortex/client.go:
```diff
@@ -284,3 +293,67 @@ func (c *Client) DeleteRuleGroup(namespace string, groupName string) error {
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return fmt.Errorf("not found")
```
`ErrNotFound` 

pracucci commented on integration/alertmanager_test.go:
```diff
@@ -40,3 +49,56 @@ func TestAlertmanager(t *testing.T) {
+	err = c.SetAlertmanagerConfig(context.Background(), cortexAlertmanagerUserConfigYaml, map[string]string{})
+	require.NoError(t, err)
+
+	time.Sleep(time.Second * 2)
```
I would prefer to rely on `time.Sleep()`. Would we get the same if we do `require.NoError(t, am.WaitSumMetrics(e2e.Equals(1), "cortex_alertmanager_configs"))`?

pracucci commented on docs/apis.md:
```diff
@@ -243,6 +243,57 @@ DELETE /api/v1/rules/{namespace}/{group_name}
+
+### Experimental API
+
+Simiarly to the Cortex Ruler, the Cortex Alertmanager supports operations using a configured object storage client as a backend for the storage and management of user's Alertmanager configuration. These API endpoints are opt-in and must be enabled via the `experimental.alertmanger.enable-api` config flag.
```
```suggestion
Similarly to the Cortex Ruler, the Cortex Alertmanager supports operations using a configured object storage client as a backend for the storage and management of user's Alertmanager configuration. These API endpoints are opt-in and must be enabled via the `-experimental.alertmanger.enable-api` CLI flag.
```

pracucci commented on pkg/alertmanager/storage.go:
```diff
@@ -8,26 +8,39 @@ import (
 	Local    local.StoreConfig `yaml:"local"`
+
+	GCS gcp.GCSConfig `yaml:"gcs"`
+	S3  aws.S3Config  `yaml:"s3,omitempty"`
```
Ups, could you remove `omitempty` here as well, please?

gotjosh commented on pkg/alertmanager/api.go:
```diff
@@ -0,0 +1,139 @@
+	cfgDesc, err := alerts.ToProto(cfg.AlertmanagerConfig, cfg.TemplateFiles, userID)
+	if err != nil {
+		level.Error(logger).Log("err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
This would never return an error, I removed the error checking entirely.

gotjosh commented on pkg/alertmanager/alerts/objectclient/store.go:
```diff
@@ -0,0 +1,96 @@
+
+// GetAlertConfig returns a specified user's alertmanager configuration
+func (a *AlertStore) GetAlertConfig(ctx context.Context, user string) (alerts.AlertConfigDesc, error) {
+	cfg, err := a.getAlertConfig(ctx, alertPrefix+user)
```
Great shout.

pracucci commented on integration/alertmanager_test.go:
```diff
@@ -40,3 +48,56 @@ func TestAlertmanager(t *testing.T) {
 	assertServiceMetricsPrefixes(t, AlertManager, alertmanager)
 }
+
+func TestAlertmanagerStoreAPI(t *testing.T) {
```
The last CI execution failed:
```
=== RUN   TestAlertmanagerStoreAPI
15:33:37 Starting minio-9000
15:33:39 Ports for container: e2e-cortex-test-minio-9000 Mapping: map[9000:32772]
15:33:39 minio-9000: Attempting encryption of all config, IAM users and policies on MinIO backend
15:33:40 Starting alertmanager
15:33:41 Ports for container: e2e-cortex-test-alertmanager Mapping: map[80:32774 9095:32773]
    TestAlertmanagerStoreAPI: alertmanager_test.go:80: 
        	Error Trace:	alertmanager_test.go:80
        	Error:      	Received unexpected error:
        	            	setting config failed, 
        	Test:       	TestAlertmanagerStoreAPI
15:33:41 Killing alertmanager
15:33:41 Killing minio-9000
15:33:41 minio-9000: Exiting on signal: TERMINATED
--- FAIL: TestAlertmanagerStoreAPI (6.15s)
```

Details: https://app.circleci.com/pipelines/github/grafana/cortex/623/workflows/50bb6d2d-158a-494e-b251-0ade70938ce1/jobs/7906/steps

Could you take a look, please?

pracucci commented:
I [pushed a commit](https://github.com/cortexproject/cortex/pull/2768/commits/ade80840be17e1ea3dad5e6249bd5d02e354fc06) to fix the local docker setup (conflicting ports and bucket folder not committed for a smooth 1st time user experience).
